lib/cli/do_volapi.js

```diff
@@ -0,0 +1,393 @@
+
+        function createVolApiSvc(ctx, next) {
+            if (ctx.volapiSvc) {
+                return next();
+            }
+
+            var domain = self.sdcadm.sdc.metadata.datacenter_name + '.' +
+                    self.sdcadm.sdc.metadata.dns_domain;
+            var svcDomain = svcData.name + '.' + domain;
+
+            self.progress('Creating "volapi" service');
+            ctx.didSomething = true;
+
+            svcData.params.image_uuid = ctx.volapiImg.uuid;
+            svcData.metadata['user-script'] = ctx.userScript;
+            svcData.metadata['SERVICE_DOMAIN'] = svcDomain;
+            svcData.params.billing_id = ctx.volapiPkg.uuid;
+            delete svcData.params.package_name;
+
+            self.sdcadm.sapi.createService('volapi', self.sdcadm.sdc.uuid,
+                    svcData, function (err, svc) {
+                if (err) {
+                    return next(new errors.SDCClientError(err, 'sapi'));
+                }
+                ctx.volapiSvc = svc;
+                self.log.info({svc: svc}, 'created volapi svc');
+                next();
+            });
+        },
+
+        /* @field ctx.headnode */
+        function getHeadnode(ctx, next) {
+            self.sdcadm.cnapi.listServers({
+                headnode: true
+            }, function (err, servers) {
+                if (err) {
+                    return next(new errors.SDCClientError(err, 'cnapi'));
+                }
+                ctx.headnode = servers[0];
+                return next();
+            });
+        },
+        function createVolApiInst(ctx, next) {
+            if (ctx.volapiInst) {
+                return next();
+            }
+
+            self.progress('Creating "volapi" instance');
+            ctx.didSomething = true;
+
+            var instOpts = {
+                params: {
+                    alias: 'volapi0',
+                    server_uuid: ctx.headnode.uuid
+                }
+            };
+            self.sdcadm.sapi.createInstance(ctx.volapiSvc.uuid, instOpts,
+                    function (err, inst) {
+                if (err) {
+                    return next(new errors.SDCClientError(err, 'sapi'));
+                }
+                self.progress('Created VM %s (%s)', inst.uuid,
+                    inst.params.alias);
+                ctx.newVolApiInst = inst;
+                next();
+            });
+        },
+        function addSharedVolumesPackages(ctx, next) {
+            function createPackageSettings(packageSize) {
+                assert.number(packageSize, 'packageSize');
+                assert.ok(packageSize > 0);
+
+                return {
+                    size: packageSize,
+                    owner_uuids: [self.sdcadm.config.ufds_admin_uuid]
+                };
+            }
+
+            var packagesSettings =
+                NFS_SHARED_VOLUMES_PKG_SIZES.map(createPackageSettings);
+
+            self.progress('Adding NFS shared volume packages');
```

Comment:
Perhaps only emit a progress message when adding each volume package. That way if re-running and the packages already exist, there is no noise.